### ISSUE TITLE
3.2 BETA 6 - Major Editor Viewport Performance Degradation

### ISSUE THREAD

**Issue by Jakob-PB** [godotengine/godot#35065]
**Godot version:**
3.2 BETA 6

**OS/device including version:**
Windows 10 Pro 1903
GPU: NVIDIA GeForce GTX 1070
Primary Display: 2560x1440 @ 144Hz
Secondary Display: 1920x1080 @ 60Hz

**Issue description:**
When using Godot 3.2 Beta 5 projects maintain at or near 144 FPS (6.94 ms) in the editor viewport while moving the camera. When opening (or creating) a project in Godot 3.2 Beta 6 the editor viewport performs at around 72 FPS (13.89 ms) while moving the camera. The problem appears in both brand new projects and project templates.

This is not an error with the FPS readout because the difference in performance of the editor is visually discernible without the aid of an FPS overlay.

**Steps to reproduce:**
1. Create a new project in Godot 3.2 Beta 6.
2. Toggle the editor viewport FPS display via the perspective menu (i.e. "Perspective->View FPS") in the viewport. 
3. Note the lower performance when compared to performing the same steps for Godot 3.2 Beta 5 or Godot 3.1.2.

Note: You may not be able to reproduce this issue without a monitor with a 120Hz or higher refresh rate.

**Comment by Jakob-PB** [comment]
I did some digging around in other beta 6 issues and found the cause of this problem. It looks like it's related to #34968 and having sync_via_compositor enabled in projects. This settings also seems to become enabled automatically for beta 5 projects at are opened in beta 6.

If you turn off `Vsync Via Compositor` under `Project Settings->Display->Window` and restart the editor then your viewport performance will be fixed.

**Comment by clayjohn** [comment]
Looks like this is a duplicate of https://github.com/godotengine/godot/issues/35038 then. Closing in favour of https://github.com/godotengine/godot/issues/35038.

**Comment by hoontee** [comment]
@Jakob-PB Does setting "Vertical sync" to "Use the 3D application setting" in the Nvidia control panel fix your issue?

**Comment by Jakob-PB** [comment]
@hoontee For me vsync was already set to "Use the 3D application setting" in the Nvidia control panel.

I tested the other vsync settings and both "on" and "adaptive" seem to help the issue while Vsync Via Compositor is on although there is a performance impact (I'm at around ~122FPS or ~8ms in an empty scene so I'm losing about 20 frames/1 millisecond to something). I disconnected my second monitor while testing this just to make sure it's not a factor.

The most optimal setting combination in terms of performance is still vsync set to "Use the 3D application setting" and Vsync Via Compositor off.

**Comment by hoontee** [comment]
@Jakob-PB Just to be clear, what settings are causing you to get halved framerate when _vsync_via_compositor_ is enabled?

**Comment by Jakob-PB** [comment]
> 
> 
> @Jakob-PB Just to be clear, what settings are causing you to get halved framerate when _vsync_via_compositor_ is enabled?

- Use the 3D application setting
- Off
- Adaptive (half refresh rate)
- Fast

**Comment by hoontee** [comment]
@Jakob-PB I've made a few fixes to the PR. Replace **platform/windows/context_gl_windows.cpp** with this file and recompile:
[context_gl_windows.zip](https://github.com/godotengine/godot/files/4061916/context_gl_windows.zip)

It will also now tell you in the console if it detects any discrepancies with the graphics driver ("VSYNC FORCED ON", "VSYNC FORCED OFF", "VSYNC FORCED OTHER"). Use this to aid in further testing.
![image](https://user-images.githubusercontent.com/5272529/72388670-2e963880-36ec-11ea-9075-384b344509ca.png)

**Comment by Jakob-PB** [comment]
@hoontee Recompiled and I'm seeing the same performance results I reported on earlier.

Not sure if this is useful information but I retested with all Nvidia vsync settings and I could never get VSYNC FORCED OFF to show up (I would assume it would show up when I force vsync off in the control panel). However, VSYNC FORCED ON and VSYNC FORCED OTHER showed up as expected for vsync set to `On` and `Adaptive` respectively.

**Comment by hoontee** [comment]
@Jakob-PB I'm perplexed.

For both of the devices I've tested (AMD & Nvidia), halved framerate only occurred when vsync was forced (On or Adaptive) in the graphics driver settings. The PR fixed it for both devices.
For you, halved framerate occurred in every situation *except* when vsync was forced (On or Adaptive) in the graphics driver settings. The PR did nothing.

The only thing that makes sense right not is VSYNC FORCED OFF not showing up - the PR doesn't detect a discrepancy because _vsync_via_compositor_ technically disables vsync. If _vsync_via_compositor_ is disabled, the message appears.

The only thing I could think of you doing at this point is going through every graphics driver setting and recording what's not set to default.

**Comment by Jakob-PB** [comment]
@hoontee Just went through the settings and looks like everything is default except "Monitor Technology" is set to "Fixed Refresh" instead of "Gsync Compatible". I restored all the Nvidia settings to default just to be sure and after a restart I'm still seeing the same behavior.

I'm on driver version 441.87 (the latest of the game ready edition) and Windows version 1903. Perhaps something changed in the new 1909 update or in a driver update?

I'm also going to try to get in touch with a friend who I know also has an Nvidia card and a 144Hz monitor.

**Comment by Jakob-PB** [comment]
@hoontee By the way, (just to be sure we're not testing different things) I am strictly talking about editor viewport performance being halved. The frame rate of the running game is locked at 144Hz for me with Vsync Via Compositor on, so from that perspective it works perfectly. Is there a difference in how the engine handles vsync for the editor vs game builds?

**Comment by hoontee** [comment]
@Jakob-PB Every time you modify a graphics driver setting, you must restart the editor to see the changes in effect. Have you been doing that?

And there's no difference between the editor and game builds from what I've noticed.

**Comment by Jakob-PB** [comment]
@hoontee Yeah I've been restarting the editor for all of this testing.

**Comment by hoontee** [comment]
@Jakob-PB Dumb question, but you're testing the game in windowed mode, correct?

If you are, I honestly have no idea where to look next.

**Comment by Jakob-PB** [comment]
@hoontee Yep, game builds are running windowed. 🙁

**Comment by hoontee** [comment]
@Jakob-PB (For context, my laptop outputs at 120Hz.)
I just checked and I can confirm that:
- Editor framerate gets reduced to ~100 FPS (not halved) when _vsync_via_compositor_ is enabled.
- The reduction does not occur in games.

This is not what the PR is meant to fix. This issue appears to be a separate bug related specifically to performance degradation from enabling _vsync_via_compositor_.

Since the PR fixes my issue with halved framerate in-game, and it doesn't affect your issue positively or negatively, I'm going to reopen #35111.

**Comment by hoontee** [comment]
@clayjohn Can you reopen this issue? It ended up not being a duplicate of #35111.

**Comment by TerminalJack** [comment]
The reduction in framerate from 120 FPS to 100 FPS is likely due to the fact that `DwmFlush()`, which is used when `vsync_via_compositor` is enabled, will always block while waiting for the vertical blank period.  On the other hand, `SwapBuffers()` with a swap interval of 1 (the ideal way to achieve vertical sync), doesn't necessarily have to block and wait for the vertical blank period  In this case the game can start processing the next frame right away.

In other words, the `vsync_via_compositor` feature should really only be used as a last-resort.
